### PR TITLE
Fix filtering of some auto-detected and customized tasks

### LINKED ISSUE #9494: Tasks: Some tasks from tasks.json are not displayed in `Task: Run Task... command`
### Bug Description:

Some VSCode tasks in tasks.json do not appear in Theia's Task: Run Task... command. Specifically it seems to occur for tasks of types that have available task providers yet do not match perfectly with an auto-provided task.

### Steps to Reproduce:

Full reproduce instructions and workspace available here:
https://github.com/bd82/theia_tasks_filter_bug

### Additional Information

- Operating System: Win10
- Theia Version: 1.13.0

### PR DESCRIPTION
**Pull request by bd82**
fixes #9494

#### What it does
Resolves issue: https://github.com/eclipse-theia/theia/issues/9494

#### How to test
A repo reproducing the issue is provided here
- https://github.com/bd82/theia_tasks_filter_bug
-  
#### Review checklist

- [ ] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/task/src/browser/task-configurations.ts [hunk 1]
```diff
@@ -129,16 +129,14 @@ export class TaskConfigurations implements Disposable {
...
-    async getTasks(token: number): Promise<TaskConfiguration[]> {
+    async getTasks(): Promise<TaskConfiguration[]> {
         const configuredTasks = Array.from(this.tasksMap.values()).reduce((acc, labelConfigMap) => acc.concat(Array.from(labelConfigMap.values())), [] as TaskConfiguration[]);
         const detectedTasksAsConfigured: TaskConfiguration[] = [];
-        for (const [rootFolder, customizations] of Array.from(this.taskCustomizationMap.entries())) {
+        for (const [, customizations] of Array.from(this.taskCustomizationMap.entries())) {
             for (const cus of customizations) {
-                // TODO: getTasksToCustomize() will ask all task providers to contribute tasks. Doing this in a loop is bad.
-                const detected = await this.providedTaskConfigurations.getTaskToCustomize(token, cus, rootFolder);
+                const detected = this.taskDefinitionRegistry.getDefinition(cus);
```
Comment (bd82) [inline packages/task/src/browser/task-configurations.ts:137]: Question to reviewers:
- Did `getTaskToCustomize` do anything else other than the incorrect filtering?
- There seemed to be some fixes related to `scopes` there...
  - https://github.com/eclipse-theia/theia/blob/92a7d05c35b43a72560a514d51d1a65be3ff5899/packages/task/src/browser/provided-task-configurations.ts#L123

File: packages/task/src/browser/task-configurations.ts [hunk 2]
```diff
@@ -54,7 +54,7 @@ export class TaskConfigurations implements Disposable {
     /**
      * Map of source (path of root folder that the task configs come from) and task customizations map.
      */
-    protected taskCustomizationMap = new Map<string, TaskCustomization[]>();
+    protected taskCustomizationMap = new Map<string, TaskConfiguration[]>();
```
Comment (bd82) [inline packages/task/src/browser/task-configurations.ts:57]: Question to reviewers:

- Is it safe to change this to `TaskConfiguration`
- See the code below, it seems the `_scope` property would always exist
  and another branch seems to assume the label property also exists.
- https://github.com/eclipse-theia/theia/pull/9501/files#diff-a4fc4b67b6c99119f65bcbfd083563d1a18a56d07410a7578d06b767c81697caR402-R406

File: packages/task/src/browser/task-configurations.ts [hunk 3]
```diff
@@ -129,16 +129,14 @@ export class TaskConfigurations implements Disposable {
...
      */
-    async getTasks(token: number): Promise<TaskConfiguration[]> {
+    async getTasks(): Promise<TaskConfiguration[]> {
         const configuredTasks = Array.from(this.tasksMap.values()).reduce((acc, labelConfigMap) => acc.concat(Array.from(labelConfigMap.values())), [] as TaskConfiguration[]);
         const detectedTasksAsConfigured: TaskConfiguration[] = [];
-        for (const [rootFolder, customizations] of Array.from(this.taskCustomizationMap.entries())) {
+        for (const [, customizations] of Array.from(this.taskCustomizationMap.entries())) {
             for (const cus of customizations) {
-                // TODO: getTasksToCustomize() will ask all task providers to contribute tasks. Doing this in a loop is bad.
-                const detected = await this.providedTaskConfigurations.getTaskToCustomize(token, cus, rootFolder);
```
Comment (bd82) [inline packages/task/src/browser/task-configurations.ts:138]: Comment to reviewers:
- There exists one more calls in the code to `getTaskToCustomize()`
- https://github.com/eclipse-theia/theia/blob/6439a6ddb30ced16fae54d12f9af22f332e8e64e/packages/task/src/browser/task-configurations.ts#L189-L204
- I do not know if that flow also has a related filtering bug or if its working as "intended".
- I'd rather not expand the scope of this PR to other bugs as I am already getting lost in the complexity here...

File: packages/task/src/browser/task-configurations.ts (hunk 1 above)
Comment (bd82) [inline packages/task/src/browser/task-configurations.ts:137]: Now when I review my code again it seems that we should just add all the tasks from the `taskCustomizationMap` because any task added to `taskCustomizationMap` already passed the check of `isDetected`  inside `reorganizeTasks()` method.

**Am I missing something?** 
What was the purpose of the original (buggy?) filtering in `getTaskToCustomize()`

Comment (tsmaeder) [reply packages/task/src/browser/task-configurations.ts:137]: I won't have time to look at this in-depth today, but getTasksToCustomize() did invoke "refreshTasks()" indirectly: so you no longer are asking the task providers to provide tasks anymore. But we need to do that once per user action (the "token" is to prevent providers being asked repeatedly.

Comment (bd82) [reply packages/task/src/browser/task-configurations.ts:137]: I will look into this again

Comment (bd82) [reply packages/task/src/browser/task-configurations.ts:137]: As far as I can see:
- `getTasksToCustomize()` was used to find the `detectedTask
- and then merge it with the customization in `getTasks()` and add the merged task to the results.
- However, it seems the `taskCustomizationMap` already has entries of `TaskConfiguration` (with _scope/label) so this merge should be a NOOP.

Although this merge may have meaning from a mutability perspective due to returning a **new** object.

File: packages/task/src/browser/task-configurations.ts [hunk 4]
```diff
@@ -401,7 +399,11 @@ export class TaskConfigurations implements Disposable {
                 }
                 const transformedTask = this.getTransformedRawTask(taskConfig, scope);
                 if (this.isDetectedTask(transformedTask)) {
-                    addCustomization(scopeKey, transformedTask);
+                    // TODO: `getTransformedRawTask` is it safe to assume `transformedTask` is a `TaskConfiguration`?
+                    //       - all branches in `getTransformedRawTask` seem to add the `_scope` property.
+                    //       - the `else` branch of this if statement seems to assume a `label` property exists.
+                    //       - These two properties are what `TaskConfiguration` adds "on top" of `TaskCustomization`...
+                    addCustomization(scopeKey, transformedTask as TaskConfiguration);
```
Comment (colin-grant-work) [inline packages/task/src/browser/task-configurations.ts:406]: If it seems like what is returned is always going to be a `TaskConfiguration`, then it would be safer to change the return type of `getTransformedRawTask` rather than to recast the return here. That way TS will help us ensure that this assumption is and remains correct.

Comment (bd82) [reply packages/task/src/browser/task-configurations.ts:406]: Hello @colin-grant-work 

I'm not 100% sure it is safe to make this assumption.
Which is why I left this `TODO` for the review discussion...

I can see that the `_scope` property will always be added inside `getTransformedRawTask()`
But I am not sure about the existence of the `label` property.

Comment (colin-grant-work) [reply packages/task/src/browser/task-configurations.ts:406]: One case in which label wouldn't be defined would be the following:  
1. Contribute a task type using Theia's internal mechanisms
2. Define a customization of that task type in a `tasks.json` without  a `label` field - maybe having no label is OK for that task type if it isn't supposed to be accessible by the task enumerating quick-picks.

It seems like that's the scenario that this code is guarding against. The assumption seems to be something like this: labels are for 'instance' task types (i.e. `TaskConfiguration`), where the user is in control of all the important paramaters, but in cases where the user only tweaks a few parameters (i.e. `TaskCustomization`), the `type` field itself is the identifier, and there _shouldn't_ be multiple instances that need their own identifiers. Those assumptions may not be correct - I'm not very familiar with the contributed task system, and this is likely something that @tsmaeder could speak to in greater detail - but I think that's what's behind the distinction here.

File: packages/task/src/browser/task-configurations.ts (hunk 1 above)
Comment (tsmaeder) [reply packages/task/src/browser/task-configurations.ts:137]: Can you guarantee that "refreshTasks" has been called at least once before this method is invoked? Because we need to ask task providers once per user interaction what their tasks are. They can be completely different for each execution of a "Run Task..." command.

Comment (bd82) [reply packages/task/src/browser/task-configurations.ts:137]: I cannot guarantee that, but we can make `refreshTasks` public and invoke it at the start of the function to
be on the safe side...

**Comment by bd82** [comment]
## Status Update

I don't think I can fully resolve this issue.
I can contribute with:
- Identifying the issue and provide a [reproducible scenario](https://github.com/eclipse-theia/theia/issues/9494)
- Identifying things that seem like possible logic bugs in the existing code and raise them for discussion.

But It seems much more **in-depth knowledge of the Tasks in Theia** and their implementation (7,000~ LOC...)
is needed to actually resolve this.

From my **limited** interaction with the `task` package. I get the feeling something here is **far too complex**.
I don't know if this is due to any, all, or none of the possible causes below:
- Legacy behaivor.
- Supporting both VSCode tasks and Theia Tasks.
- Patches on Patches.
- lack of tests.
- Some inherent design issues.

Some things I do recognize are:
- There are almost 7,000 LOC of code in the task `src` directory, only ~300 of which are tests.
  - Which seems like a-lot of code.
- There are 49 open issues and 117 closed issues labeled under `tasks`. 
  - Which seem like a-lot of related issues.

If this was a project I "owned" (and knew in-depth) I would try to evaluate the reasons for the complexity
and the feasibility of a re-write. But that is far beyond the scope of my capabilities here...

And perhaps my gut feeling is completely wrong and all the complexity here is fully inherent to the problem space (no accidental complexity...).

### So to summarize:
**Is there anyone from Theia who understands this area deeply and can attempt to resolve the original issue https://github.com/eclipse-theia/theia/issues/9494 ?**

File: packages/task/src/browser/task-configurations.ts (hunk 1 above)
Comment (tsmaeder) [reply packages/task/src/browser/task-configurations.ts:137]: Calling refreshTasks() with a token is a no-op if it has been called before with the same token, so this should be safe.

**Comment by tsmaeder** [comment]
@bd82 thanks a lot for working on this. The tasks system in Theia is quite the beast that suffers (IMO) from getting some of the basic concepts wrong and then working around that (and don't get me started on the naming of things ;-)). Simplifying things (like your PR does) is a step in the right direction. Do not get discouraged!

**Comment by colin-grant-work** [comment]
> * There are almost 7,000 LOC of code in the task `src` directory, only ~300 of which are tests.
>   
>   * Which seems like a-lot of code.

I completely agree with this sentiment, and if you wanted to shift the focus of this PR from implementing a particular solution to increasing test coverage, I would support that. That would quickly place you among the most knowledgeable developers in this area and almost certainly expose a number of oddities and conflicting assumptions that would provide good starting points for later fixes and simplifications.

**Comment by bd82** [comment]
> I completely agree with this sentiment, and if you wanted to shift the focus of this PR from implementing a particular solution to increasing test coverage, I would support that.

I am trying to solve the original bug as it affects a business scenario for the organization I work for...

> Simplifying things (like your PR does) is a step in the right direction. Do not get discouraged!

I have limited time before I go on a long summer vacation.
I will attempt to investigate the `label` question before I leave for vacation,
 as I think that is the main open question in this PR
- Is it mandatory? 
  - In Theia.
  - In VSCode.
- Could the deleted code in this PR have added a "missing" label in some scenarios?

**Comment by bd82** [comment]
Okay what I've learned is:
- This `label` property is used extensively in VSCode as well.
- It is not part of any of the tasks related APIs I can find on VSCode official API docs.
- It seems to be a computed property.
- I cannot seem to find where in fact it is computed though.

Currently this makes me think that `getTasks()` should perhaps `TaskCustomization[]`  instead of `TaskConfiguration[]`.
But this depends on what happens to plain tasks written in the tasks.json
Do they undergo some transformation and have a `label` property added to them in the internal in memory representation
or is `getTasks()` expected to return those tasks.json tasks "as is" without any changes.

But I need to investigate it farther, but that won't happen anytime soon as I will be OOO for most of the summer... 😄

**Comment by JonasHelming** [comment]
@tsmaeder : Could you please decide on what to do with this PR?

**Comment by tsmaeder** [comment]
Since this would need some work which no-one can do right now, let's close the PR.
